### PR TITLE
15576: Invoke local residency service (Local Hub)

### PR DESCRIPTION
**Pull request by nisanthyaganti9**
|Field|Value|
|-----|-----|
| Redmine Ticket Number | [15576](https://devops.dchbx.org/redmine/issues/15576) |
| App-Dev Road Map # | n/a |
| Start Date | n/a |
| Due Date | n/a  |
| App-Dev Priority | UP-3 |
| Development Story Points | 2 |
| Peer Review Story Points | n/a |
| Ticket Author | @TreyE |

### CODE REVIEW

File: app/controllers/concerns/vlp_doc.rb [hunk 1]
```diff
@@ -52,4 +52,12 @@ def get_vlp_doc_subject_by_consumer_role(consumer_role)
     docs_for_status_uploaded = consumer_role.vlp_documents.where(:subject=>{"$in" => docs_for_status})
     docs_for_status_uploaded.any? ? docs_for_status_uploaded.order_by(:updated_at => 'desc').first.subject : nil
   end
+
+  def can_retrigger_residency?(no_dc_address, fm_person=nil)
+    @person = fm_person if fm_person.present?
+    @person.age_on(TimeKeeper.date_of_record) > 18 &&
+    @person.no_dc_address == true &&
+    no_dc_address == "false" &&
+    @family.active_household.hbx_enrollments.where(:aasm_state.in => HbxEnrollment::ENROLLED_STATUSES).flat_map(&:hbx_enrollment_members).flat_map(&:family_member).flat_map(&:person).include?(@person)
```
Comment (lisyk) [inline app/controllers/concerns/vlp_doc.rb]: this is the good check. let's combine it with existing check_for_critical_changes method

File: app/controllers/concerns/vlp_doc.rb [hunk 2]
```diff
@@ -52,4 +52,12 @@ def get_vlp_doc_subject_by_consumer_role(consumer_role)
     docs_for_status_uploaded = consumer_role.vlp_documents.where(:subject=>{"$in" => docs_for_status})
     docs_for_status_uploaded.any? ? docs_for_status_uploaded.order_by(:updated_at => 'desc').first.subject : nil
   end
+
+  def can_retrigger_residency?(no_dc_address, fm_person=nil)
+    @person = fm_person if fm_person.present?
+    @person.age_on(TimeKeeper.date_of_record) > 18 &&
+    @person.no_dc_address == true &&
```
[the remark this answers is not in the document]
Comment (nisanthyaganti9) [reply app/controllers/concerns/vlp_doc.rb]: "@person.no_dc_address" denotes current status & "no_dc_address" is the params value representing new change. we only need to trigger hub if - the change in address is from "no_dc_address" to "dc_address".

File: app/controllers/insured/family_members_controller.rb [hunk 3]
```diff
@@ -127,6 +127,7 @@ def update
     end
     consumer_role = @dependent.family_member.try(:person).try(:consumer_role)
     consumer_role.check_for_critical_changes(params[:dependent]) if consumer_role
+    @person.consumer_role.retrigger_residency! if @person.consumer_role.present? && can_retrigger_residency?(params["dependent"]["no_dc_address"], @dependent.family_member.person)
```
[the remark this answers is not in the document]
Comment (nisanthyaganti9) [reply app/controllers/insured/family_members_controller.rb]: as per Sarah we have to re-trigger hub whenever address change happened from no-dc-address to dc-address. & "can_retrigger_residency?" only returns true for this scenario.

File: app/models/consumer_role.rb [hunk 4]
```diff
@@ -468,6 +489,7 @@ def can_receive_electronic_communication?
   end
 
   def invoke_verification!(*args)
+    start_residency_verification_process unless person.no_dc_address
```
[the remark this answers is not in the document]
Comment (nisanthyaganti9) [reply app/models/consumer_role.rb]: I am not using this. This appeared b'coz of  merging "residency_verification" branch into this & we already discussed about this the other day while reviewing #15581.

File: app/controllers/concerns/vlp_doc.rb (hunk 1 above)
Comment (nisanthyaganti9) [reply app/controllers/concerns/vlp_doc.rb]: okay

**Pull request merged**